### PR TITLE
:seedling: Convert stakeholder table to new table abstraction

### PR DESCRIPTION
**Pull request by ibolton336**
- useTableControls implementation for Stakeholders expandable form 
@mturley Nice work! We don't have to necessarily merge this as I know it is still under construction a bit. Just wanted to explore the changes. Works flawlessly. Only thing I did notice - we might need a way to make individual rows expandable. If I have a stakeholder with no groups assigned, I wonder if there is a way to make only that row not have an expandable icon displayed without shifting the columns/table rows.

### CODE REVIEW

**Comment by mturley** [comment]
Nice!

Interesting, yeah, I think we can probably take advantage of the composable nature there and have some conditional expression on that cell where you only spread the `getCompoundExpandTdProps(...)` if that row should have it, else spread `getTdProps(...)`. Either that or we can add an `isCellExpandable` boolean to `getCompoundExpandTdProps` and have it exclude the necessary props when that's false.

**Review by mturley — DISMISSED** [review]
LGTM! I'm totally fine merging this if you want, I already keep the migration waves / jira tables up to date whenever I change these hooks so it's not too much effort to manage one more. Within reason, the more tables we have using it the more confident we can be in the abstraction

**Comment by mturley** [comment]
Actually @ibolton336 I approved too quickly, some minor things:
* On main, it looks like there is padding/margin around the contents of the expanded row content, and here the text is right up against the borders there
* For this single-expand use case I actually wanted to add support for automatically rendering that expand toggle cell as part of `<TableRowContentWithControls>`, but we can do that as a separate refactor after this if you want. Would just mean a little less code in the table.

**Comment by ibolton336** [comment]
> Actually @ibolton336 I approved too quickly, some minor things:
> 
> * On main, it looks like there is padding/margin around the contents of the expanded row content, and here the text is right up against the borders there

For some reason I am seeing an extra <td> in the expandable row that seems to be implicitly created by the legacy table.  

Screenshot 2023-04-20 at 9 26 13 AM

File: client/src/app/pages/controls/stakeholders/stakeholders.tsx [hunk 1]
```diff
@@ -284,16 +285,7 @@ export const Stakeholders: React.FC = () => {
...
-                            rowIndex,
-                            isExpanded: isCellExpanded(stakeholder),
-                            onToggle: () =>
-                              setCellExpanded({
-                                item: stakeholder,
-                                isExpanding: !isCellExpanded(stakeholder),
-                              }),
-                            expandId: "composable-expandable-example",
-                          }}
+                          {...getExpandTdProps({ item: stakeholder, rowIndex })}
```
Comment (mturley) [inline client/src/app/pages/controls/stakeholders/stakeholders.tsx:288]: Can we rename it to `getSingleExpandTdProps` to distinguish it from the compound expand version?

File: client/src/app/pages/applications/components/application-form/application-form.tsx [hunk 2]
```diff
@@ -666,18 +666,20 @@ export const ApplicationForm: React.FC<ApplicationFormProps> = ({
...
-                aria-label="binary packaging details"
-                bodyContent={t("message.binaryPackaging")}
-                className="popover"
-              >
-                <span className="pf-c-icon pf-m-info">
-                  <QuestionCircleIcon />
-                </span>
-              </Popover>
-            }
+            formGroupProps={{
```
Comment (mturley) [inline client/src/app/pages/applications/components/application-form/application-form.tsx:669]: Looks like the diff contains some stuff from another change?

File: client/src/app/pages/controls/stakeholders/stakeholders.tsx [hunk 3]
```diff
@@ -306,68 +116,244 @@ export const Stakeholders: React.FC = () => {
...
+                          <AppTableActionButtons
+                            onEdit={() => editRow(stakeholder)}
+                            onDelete={() => deleteRow(stakeholder)}
+                          />
+                        </Td>
+                      </TableRowContentWithControls>
+                    </Tr>
+                    {isCellExpanded(stakeholder) ? (
+                      <Tr isExpanded>
+                        <Td />
```
Comment (mturley) [inline client/src/app/pages/controls/stakeholders/stakeholders.tsx:322]: Oh I see, this is to space it out so it is indented past the toggle. I guess that makes sense... hm, it does kinda suck that we have to remember to put this extra Td in though. I wonder if it's worth adding a new component for `<ExpandedContentRowWithControls>` or something that renders the Tr, the two Tds with the right props, and then passes children in.

Probably not worth it for now, I'll make a note to maybe mess with that later.

File: client/src/app/shared/hooks/use-table-controls/useTableControlProps.ts [hunk 4]
```diff
@@ -173,6 +173,25 @@ export const useTableControlProps = <
...
+  }): Omit<TdProps, "ref"> => ({
+    expand: {
+      rowIndex,
+      isExpanded: isCellExpanded(item),
+      onToggle: () =>
+        setCellExpanded({
+          item,
+          isExpanding: !isCellExpanded(item),
+        }),
+      expandId: "expandable-example",
```
Comment (mturley) [inline client/src/app/shared/hooks/use-table-controls/useTableControlProps.ts:191]: These ids need to be unique for QE tests. We probably want to be able to pass in a custom id here, or maybe have it be something built-in like:
```ts
expandId: `expandable-row-${item.name}`
```

**Review by mturley — CHANGES_REQUESTED** [review]
Looking good! A couple minor things

**Comment by ibolton336** [comment]
> Actually @ibolton336 I approved too quickly, some minor things:
> 
> * On main, it looks like there is padding/margin around the contents of the expanded row content, and here the text is right up against the borders there
> * For this single-expand use case I actually wanted to add support for automatically rendering that expand toggle cell as part of `<TableRowContentWithControls>`, but we can do that as a separate refactor after this if you want. Would just mean a little less code in the table.

Thanks @mturley For all the reviews. I think I hit all of the requested updates. Whenever you have time.

**Review by mturley — APPROVED** [review]

**Comment by sshveta** [comment]
HI @mturley @ibolton336 .
In this PR the stakeholder table ID is changed to stakeholders-table .

All the other tables (like stakeholder Group , Business Service , applications page ) in MTA are using "table[aria-label='main-table']";.

Is it possible to keep the ID same and generic for all the tables ? 
That was enabling us to write common methods to delete all entities in page . 

Please let me know what you think .

**Comment by mturley** [comment]
Hi @sshveta. 

I think we failed to properly communicate here, sorry about that. As established by #443 (which perhaps should not be closed), we do want to align on using `id` for QE tests where possible (or more specifically, avoiding having tests depend on `aria-label` which is user-facing text that is subject to change).

Perhaps we should use `id="main-table"` for each of the primary tables on the page, that is a decent idea. I'm curious about @ibolton336 's thoughts.

**Comment by ibolton336** [comment]
@mturley @sshveta Sorry for the oversight here. We can certainly move to a common id for these tables if that would be helpful.

**Comment by sshveta** [comment]
Thanks @ibolton336 . Please let me know if I need to open a bug/issue for this .

**Pull request merged**